src/app/snapp_test_transaction/snapp_test_transaction.ml

```diff
@@ -0,0 +1,1247 @@
+           (Currency.Fee.to_formatted_string min_fee))
+      (Param.optional txn_fee)
+
+  let snapp_account_key =
+    Param.flag "--snapp-account-key"
+      ~doc:"PUBLIC KEY Base58 encoded public key of the new snapp account"
+      Param.(required public_key_compressed)
+
+  let amount =
+    Param.flag "--receiver-amount" ~doc:"NN Receiver amount in Mina"
+      (Param.required txn_amount)
+
+  let nonce =
+    Param.flag "--nonce" ~doc:"NN Nonce of the fee payer account"
+      Param.(required txn_nonce)
+
+  let snapp_state =
+    Param.flag "--snapp-state"
+      ~doc:
+        "FIELDS A list of 8 values that can be Integers, arbitrarty strings,  \
+         hashes, or field elements"
+      Param.(required txn_nonce)
+
+  let common_flags =
+    Command.(
+      let open Let_syntax in
+      let%map keyfile =
+        Param.flag "--fee-payer-key"
+          ~doc:
+            "KEYFILE Private key file for the fee payer of the transaction \
+             (should already be in the ledger)"
+          Param.(required string)
+      and fee = fee
+      and nonce = nonce
+      and memo = memo
+      and debug =
+        Param.flag "--debug" Param.no_arg
+          ~doc:"Debug mode, generates transaction snark"
+      in
+      (keyfile, fee, nonce, memo, debug))
+end
+
+module App_state = struct
+  type t = Snark_params.Tick.Field.t
+
+  let of_string str : t Snapp_basic.Set_or_keep.t =
+    match str with
+    | "" ->
+        Snapp_basic.Set_or_keep.Keep
+    | _ -> (
+        match
+          Or_error.try_with (fun () -> Snark_params.Tick.Field.of_string str)
+        with
+        | Ok f ->
+            Snapp_basic.Set_or_keep.Set f
+        | Error e1 -> (
+            match Signed_command_memo.create_from_string str with
+            | Ok d ->
+                let s =
+                  Random_oracle.(
+                    hash ~init:Hash_prefix.snapp_test
+                      ( Signed_command_memo.to_bits d
+                      |> Random_oracle_input.bitstring |> pack_input ))
+                in
+                Snapp_basic.Set_or_keep.Set s
+            | Error e2 ->
+                failwith
+                  (sprintf
+                     "Neither a field element nor limited length string Errors \
+                      (%s, %s)"
+                     (Error.to_string_hum e1) (Error.to_string_hum e2)) ) )
+end
+
+module Events = struct
+  type t = Snark_params.Tick.Field.t
+
+  let of_string_array (arr : string Array.t) =
+    Array.map arr ~f:(fun x ->
+        match x with
+        | "" ->
+            Snark_params.Tick.Field.zero
+        | _ -> (
+            match
+              Or_error.try_with (fun () -> Snark_params.Tick.Field.of_string x)
+            with
+            | Ok f ->
+                f
+            | Error e1 -> (
+                match Signed_command_memo.create_from_string x with
+                | Ok d ->
+                    Random_oracle.(
+                      hash ~init:Hash_prefix.snapp_test
```

Comment:
Is there any documentation on when we use this hashing? I would think that it's more likely to conceal errors than to ever be useful, and would prefer to see it removed.

In particular, a snapp developer will pretty much never use string hashing -- even we only ever use the unhashed memo outside of the circuit -- and so if this is the desired behaviour we probably want a different way of constructing a hash. A 'random field' version of the sequence state would seem to make more sense, if we'd prefer not to manually construct the field elements.

Reply:
@mrmr1993 Should we just fail here if the string does not represent a field, then?

Reply:
I think so, yes